### ISSUE TITLE
Invalid band names when merging cubes

### ISSUE THREAD

**Issue by jonathom** [Open-EO/openeo-python-client#212]
I am trying to merge two separate collections of imagery. I can't seem to access the additional band(s) afterwards though, a `ValueError: Invalid band name/index 'VV'. Valid names: ['B04', 'B08']` is thrown, see code below:
```
import openeo
con = openeo.connect("https://openeo.vito.be")
con.authenticate_basic("name", "pwd")
urk = {"west": 5.5661, "south": 52.6457, "east": 5.7298, "north": 52.7335}
t = ["2021-04-01", "2021-04-30"]

# load first datacube
cube_s2 = con.load_collection(
    "SENTINEL2_L2A_SENTINELHUB",
    spatial_extent = urk,
    temporal_extent = t,
    bands = ["B04", "B08"]
)

# second datacube
cube_s1 = con.load_collection(
    "S1_GRD_SIGMA0_ASCENDING",
    spatial_extent=urk,
    temporal_extent=t,
    bands=["VV", "VH"]
)

# resample and reduce
cube_s1 = cube_s1.resample_cube_spatial(cube_s2) # resample before mean_time
cube_s1 = cube_s1.mean_time()
cube_s2 = cube_s2.mean_time()

cube = cube_s2.merge_cubes(cube_s1)

B04 = cube.band("B04")
B08 = cube.band("B08")
VV = cube.band("VV")

graph = VV.to_json()

with open("merge_cubes_graph.txt", "w") as tfile:
    tfile.write(graph)
```
`ValueError: Invalid band name/index 'VV'. Valid names: ['B04', 'B08']`
So seems to me that band `VV` is not accessible per name after merging. Upon submission (of `cube`) to the backend, all 3 merged bands are returned, so that seems fine.

**Comment by soxofaan** [comment]
For completeness: the error happens client side before any backend-side processing:

```
<ipython-input-2-0a9a4207b09e> in <module>
     31 B08 = cube.band("B08")
---> 32 VV = cube.band("VV")

~/src/openeo/openeo-python-client/openeo/rest/datacube.py in band(self, band)
--> 435         band_index = self.metadata.get_band_index(band)

~/src/openeo/openeo-python-client/openeo/metadata.py in get_band_index(self, band)
    344     def get_band_index(self, band: Union[int, str]) -> int:
--> 345         return self.band_dimension.band_index(band)

~/src/openeo/openeo-python-client/openeo/metadata.py in band_index(self, band)
--> 116         raise ValueError("Invalid band name/index {b!r}. Valid names: {n!r}".format(b=band, n=band_names))

ValueError: Invalid band name/index 'VV'. Valid names: ['B04', 'B08']
```

**Comment by soxofaan** [comment]
The problem here with that `cube.band("VV")`  call is that the client should know what bands this merged cube has, which is not trivial because it requires the client to simulate what the backend will be doing.
It is probably feasible to get things working for this `merge_cubes` use case,  but this raises the general question if the client should  do this kind of simulation fully client side and "pre-flight".

**Comment by jonathom** [comment]
Maybe some context: The thing I'm trying to achieve here is to mask a cube with data from another cube (in opposition to the tutorials, where data from the same cube is used for masking). E.g. set all pixels with NDVI > 0.3 to NA in a Radar image.

When I'm doing that without merging the two cubes, I get a `geotrellis.layer.SpatialKey cannot be cast to geotrellis.layer.SpaceTimeKey` error although having reduced the time dimension in both cubes, so I figured maybe merging cubes might help, I then ran into this. Do you have a proposed workflow for the task above?

> the general question if the client should do this kind of simulation fully client side and "pre-flight".

I think I only partly understand where you're coming from. Because I'd think functions do have expected behaviours - I don't see where this is different from the promise of returning a datacube which is taken as granted by every following function. Could you elaborate?

**Comment by soxofaan** [comment]
> When I'm doing that without merging the two cubes, I get a ...

could you submit that issue under Open-EO/openeo-geopyspark-driver so we can address that properl?.At first sight it seems what you are trying should be possible

>  Could you elaborate?

When you do  `cube.band("B04")` (what we call "band math" functionality), the client translates this in a `reduce_dimension` process using an `array_element` with the (integer) index of that particular band "B04". That works directly after a `load_collection` because the client can inspect the metadata of the collection you loaded and fetch the correct index of that band there. 
But after a `merge_cubes` that is a lot harder: the client has,  in principle, no metadata of the result of `merge_cubes` because it has not happened yet for real on the backend. Note that your code is only executed on the backend when you submit the process graph through   `download`, `execute` or something like that. A workaround is that the client tries to simulate what the backend will be doing and guess what the metadata will be, but that is not trivial or without risk of making wrong assumptions.

For your use case a quick workaround could be to use integers instead of string in the `.band()` calls here:
```python
cube = cube_s2.merge_cubes(cube_s1)

B04 = cube.band(0)
B08 = cube.band(1)
VV = cube.band(2)
```
but that requires that you carefully determined the correct band indices yourself

**Comment by jonathom** [comment]
Thank you for clarifying:) I understand now that keeping track of and anticipating processing steps is merely possible, also with looking at the extensive `merge_cube` description examples. However,
> cube = cube_s2.merge_cubes(cube_s1)
B04 = cube.band(0)
B08 = cube.band(1)
VV = cube.band(2)

fails with the same error. Which kind of makes sense to me: If this would work, the client would need to assume that three bands come out of the process - how would that be done given the `merge_cube` complexity / why not assume their labels as well then?

**Comment by soxofaan** [comment]
> fails with the same error.

hmm indeed, there is still a check happening about the integer index being out of bound: the client at the moment copies the metadata from `cube_s2` for the metadata of the merged cube and consequently thinks the merged cube has only 2 bands, so `index=2` is out of range.

Next workaround I can think of is building `VV` as follows:
```python
VV = cube.band(1)
VV._pg.arguments["reducer"]["process_graph"].arguments["index"] = 2
```
So you build `VV` like it was the second band, but then by messing with the internals, you set the index to 2 (third band).
Warning: only do this for testing/proof of concept, do not do this in production code

**Comment by jonathom** [comment]
Thank you for the workaround, although I would only need this for production code.

Another note: Both R and JS clients submit this process to the back-end without issues, which makes me think checks are passed and this type of operation and predicting merged_cube properties works.

**Comment by soxofaan** [comment]
> Both R and JS clients submit this process to the back-end without issues, which makes me think checks are passed and this type of operation and predicting merged_cube properties works.

can you paste the those working R and JS code snippets here as reference?

**Comment by jonathom** [comment]
sure. Not the most useful operations though.

JS
```
// load module
const { OpenEO, Formula } = require('@openeo/js-client');

// put all code in an async so we can catch errors
async function example() {

    // connect to backend
    con = await OpenEO.connect("https://openeo.vito.be");
    
    // autheticate
    await con.authenticateBasic("", "");

    // get process builder help
    var builder = await con.buildProcess();

    // make spatial and temporal extent
    let urk = {"west": 5.6061, "south": 52.6957, "east": 5.7298, "north": 52.7335};

    let t = ["2021-04-15", "2021-04-30"];   

    // load first cube
    var cube_s2 = builder.load_collection(
        "SENTINEL2_L2A_SENTINELHUB",
        urk,
        t,
        ["B02", "B04", "B08"]
    );

    var cube_s1 = builder.load_collection(
        "S1_GRD_SIGMA0_ASCENDING",
        urk,
        t,
        ["VV"]
    );

    var cubes = builder.merge_cubes(cube_s1, cube_s2)

    var red = builder.filter_bands(cubes, ["B04"])
    var VV = builder.filter_bands(cubes, ["VV"])

    var merged = builder.merge_cubes(red, VV)

    // save as CSV
    result = builder.save_result(merged, "GTiff");

    // send job to backend
    var job = await con.createJob(result, "test_merged_behaviour_JS_bandname_as_array");

}

example().catch(error => console.error(error));
```

R
```
# load module
library(openeo)

# establish connection
con <- connect(host = "https://openeo.vito.be")

# authenticate
login(login_type="basic", user="", password="")

# get a process graph builder, see ?processes
p <- processes()

# create variables for loading collection
urk <- list(west = 5.5661, south = 52.6457, east = 5.7298, north = 52.7335)

t <- c("2021-04-16", "2021-04-29")

# load first datacube
cube_s2 <- p$load_collection(
  id = "SENTINEL2_L2A_SENTINELHUB",
  spatial_extent = urk,
  temporal_extent = t,
  bands=c("B02", "B04", "B08")
)

# load first datacube
cube_s1 <- p$load_collection(
  id = "S1_GRD_SIGMA0_ASCENDING",
  spatial_extent = urk,
  temporal_extent = t,
  bands=c("VV")
)

cubes <- p$merge_cubes(cube_s1, cube_s2)

red <- p$filter_bands(data = cubes, bands = "B04")
vv <- p$filter_bands(data = cubes, bands = "VV")

cube <- p$merge_cubes(red, vv)

res <- p$save_result(cube, format = "GTiff")

graph <- as(res, "Graph")
graph$validate()

# send job to back-end
job <- create_job(graph = res, title = "check_merge_handling_R")
```

**Comment by m-mohr** [comment]
> It is probably feasible to get things working for this `merge_cubes` use case, but this raises the general question if the client should do this kind of simulation fully client side and "pre-flight".

I doubt this is a good idea as it makes things really complex client-side. It's probably not just merge_cubes that would bring up such issues. But if the client does such assumptions then it not only work in some cases.

**Comment by soxofaan** [comment]
> Why is that actually happening? 

Because the VITO backend does not support labels in `array_element` yet: Open-EO/openeo-geopyspark-driver#43 . The label argument in `array_element` only came in the 1.0 branch of the api/processes if I'm not mistaken and we implemented this client-side label-to-index conversion before to that. Because that works in most cases, the priority to add label support is not quite high.
 
That being said, adding proper `array_element` label support it in the VITO backend  will indeed be a better solution compared to any form of client side simulation.


> there are a lot of cases where things will surprisingly fail (from a user POV).

what do you mean with that?

**Comment by m-mohr** [comment]
> > there are a lot of cases where things will surprisingly fail (from a user POV).
> 
> what do you mean with that?

As far as I understand, all cases where some kind of band label manipulation has taken place after load_collection will fail in the Python client (e.g. rename_labels, merge_cubes, apply_dimension, add_dimension, ...), which is a "surprise" for the user.

**Comment by soxofaan** [comment]
> As far as I understand, all cases where some kind of band label manipulation has taken place after load_collection will fail in the Python client (e.g. rename_labels, merge_cubes, apply_dimension, add_dimension, ...), 

FYI: It's only  `cube.band(label)`  calls that will have issues when there is some kind of manipulation of the band dimension  (like `merge_cubes`) between the `load_collection` and that `cube.band()` call. If there is no `cube.band()` call after, say, `merge_cubes` there is no problem.

**Comment by m-mohr** [comment]
Yes, I'm aware of that.

**Comment by soxofaan** [comment]
@jonathom   thanks for the R and JS snippents. The process you implement there is actually a bit different than what you did in your Python snippet: `filter_bands()` (even with a single band)   is technically different from a  `.band()` call (because it does a `reduce_dimension`+`array_element`  instead of just `filter_bands`).

So you can already implement those R/JS approaches in Python as follows:
```python

cubes = cube_s2.merge_cubes(cube_s1)

B04 = cubes.filter_bands(["B04"])   # instead of cube.band("B04")
VV = cubes.filter_bands(["VV"])     # instead of cube.band("VV")

merged = B04.merge_cubes(VV)
```

**Comment by jonathom** [comment]
Unfortunately, still the same error. I'll post the complete Python code here aswell:

```
# import the openeo module
import openeo

# establish connection
con = openeo.connect("https://openeo.vito.be")

# authenticate
con.authenticate_basic("", "")

# make dictionary, containing bounding box
urk = {"west": 5.5661, "south": 52.6457, "east": 5.7298, "north": 52.7335}
# make list, containing the temporal interval
t = ["2021-04-01", "2021-04-30"]

# load first datacube
cube_s2 = con.load_collection(
    "SENTINEL2_L2A_SENTINELHUB",
    spatial_extent = urk,
    temporal_extent = t,
    bands = ["B04", "B08"]
)

# mask
cube_s1 = con.load_collection(
    "S1_GRD_SIGMA0_ASCENDING",
    spatial_extent=urk,
    temporal_extent=t,
    bands=["VV", "VH"]
)

cube = cube_s2.merge_cubes(cube_s1)

B04 = cube.filter_bands(["B04"])
VV = cube.filter_bands(["VV"])

graph = VV.to_json()

with open("python_mask_2cubes.txt", "w") as tfile:
    tfile.write(graph)
```

**Comment by soxofaan** [comment]
hmm ok, so the python client is still being too "helpful" with incomplete information

another workaround that should work with current client is this after the `cube = cube_s2.merge_cubes(cube_s1)`:

```python

from openeo.metadata import Band
cube.metadata.band_dimension.bands += [Band("VV", None, None), Band("VH", None, None)]
```

**Comment by soxofaan** [comment]
I believe this has been addressed in a9239a8d818e9c52fdedad76967ae8f277a6a62a (part of recent release 0.12.0)
